### ISSUE TITLE
Money not properly representing in OSPOS

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#2748]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [x] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.2
- OSPOS git commit hash is: 
- PHP version is: 7.3
- MySQL or MariaDB version is: 5.7.26
- OS and version is: FreeBSD 11.2
- WebServer is: Apache 2.4
- Selected language is: English
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### Issue

I'm having issues with how we are currently storing and using money.  The generated JSON from my CLCdesq 3rd-Party integration is giving me:
`		"Price": 2.2200000000000002,`

instead of 2.22, I'm getting a wonky value.

In the database we have Money fields setup as a DECIMAL(15,2).  According to most of my research the best practice is either storing it as an INT, then dividing by 100 (or 10000 if storing 1/100th of a cent) or DECIMAL(19,4) or DECIMAL(22,2) (https://stackoverflow.com/questions/13030368/best-data-type-to-store-money-values-in-mysql).  The last of which is close to what we are doing.

The problem I am having is not as much with MySQL and instead with PHP.  In my CLCdesq 3rd-party integration I have the line:
`'Price' 				=> $data['unit_price'],`

in an associative array.  $data is put together and it's source is the new item view which gets processed by the save function of the Items Controller as:
`parse_decimals($this->input->post('unit_price'))`

parse_decimals is in the locale_helper and it's most likely here that we have a problem:
`function parse_decimals($number)`
`{`
`// ignore empty strings and return`
`	if(empty($number))`
`	{`
`		return $number;`
`	}`
``
`	$config = get_instance()->config;`
`	$fmt = new \NumberFormatter($config->item('number_locale'), \NumberFormatter::DECIMAL);`
``
`	$fmt->setAttribute(\NumberFormatter::FRACTION_DIGITS, $config->item('currency_decimals'));`
``
`	if(empty($config->item('thousands_separator')))`
`	{`
`		$fmt->setAttribute(\NumberFormatter::GROUPING_SEPARATOR_SYMBOL, '');`
`	}`
``
`	try`
`	{`
`		return $fmt->parse($number);`
`	}`
`	catch(Exception $e)`
`	{`
`		return FALSE;`
`	}`
`}`

The band-aid is me rounding the 3rd-party integration level.  This is not a good idea because the potential for issues in other parts of the code (such as tax calculation, report generation, etc).  It sounds like the most surefire would be to overhaul the code to storing everything as an integer, but then dividing by 100 at display and calculation time.  That's potentially a lot of work.  Perhaps something I'm missing here.  Thoughts?

**Comment by daN4cat** [comment]
I was under the impression that we stored values with more than 2 decimals. I believe we had in the past cases for 3 decimals.

**Comment by daN4cat** [comment]
BTW the problem with storing it as integer is that if you move from 2 to 3 decimals due to some new regulatory obligation you would have old data x 100 and new data x 1000.
How are you going to manage that with an integer story?
Where having a float with 3 or 4 decimals will not be a problem for you.

**Comment by objecttothis** [comment]
> 
> 
> BTW the problem with storing it as integer is that if you move from 2 to 3 decimals due to some new regulatory obligation you would have old data x 100 and new data x 1000.
> How are you going to manage that with an integer story?
> Where having a float with 3 or 4 decimals will not be a problem for you.

I'm not married to the idea of storing it as an integer.  I think that is the limitation of storing as an integer.  Although, if you change the number of required decimals, the old data is easy to change... you just run a migration script that would multiply all money fields by 10 (if increasing the number of decimals).  Decreasing would require you to either truncate or round (going from 4 to 3, for example).

**Comment by objecttothis** [comment]
> 
> 
> I was under the impression that we stored values with more than 2 decimals. I believe we had in the past cases for 3 decimals.

My database says 15,2 for unit_price in the items table.  What does yours say?  Mine was migrated from an old 1.x version, so maybe I missed some boat somewhere.

At any rate, I'm getting a wonky value and I'm guessing that its endemic of the system not just my code since my code is using our locale_helper to get it's values.  The question is, what do we do about it?

**Comment by SteveIreland** [comment]
unit_price has been stored as 2 decimals for years.  If you take it out of the database as a 2 decimal position value and then through some JSON transformation a rounding issue manifests itself then it seems logical that this is a but in the transformation code and not a database issue.

If you run a simple SQL against the data does it display something other than 2 decimal positions?  That would be really weird.

**Comment by objecttothis** [comment]
@SteveIreland I don't think the database is the issue.  I think it's PHP that is the issue.  There are dozens of articles about PHP and other languages talking about how they handle floating point values and how to "tame the beast".

Honestly, I'd prefer the least invasive change that gets the job done.  No need to overhaul the code if we can help it.  We can also make a mistake if we ignore it as if it's not an issue.

**Comment by jekkos** [comment]
@objecttothis how did you get to this value? I think it should be easy now to add a unit test that verifies this behavior in our code. I would be surprised that if we set formatting to use the configured amount of decimals, it ends up with a lot more.

Issue might be somewhere after in generating the json (yet another way of representing numbers there).

**Comment by objecttothis** [comment]
@SteveIreland here is an example that you asked for:
![database values](https://user-images.githubusercontent.com/17935339/77162210-db15e180-6ac4-11ea-8d9a-e1c8eab1d9dd.jpg)

and here is the JSON that my 3rd party integration is spitting out:
![JSON](https://user-images.githubusercontent.com/17935339/77162959-8ffcce00-6ac6-11ea-9b6b-e7fa5bae23b7.jpg)

The code in the items controller is `'unit_price' => parse_decimals($this->input->post('unit_price')),` which is getting rounded to the nearest cent when it gets sent to the database.  The field in the POST sent by the view was entered as "16.99"

**Comment by jekkos** [comment]
@objecttothis Ok this is really weird, as I showed in the unit test, the decimal rounding works as it should. Can you check the code in between the parse en the display? Maybe extend the unit test to see where the formatting changes.

**Comment by objecttothis** [comment]
> 
> 
> @objecttothis Ok this is really weird, as I showed in the unit test, the decimal rounding works as it should. Can you check the code in between the parse en the display? Maybe extend the unit test to see where the formatting changes.

@jekkos, my code isn't manipulating the value, it's just passing it through as is until the json_encode.  I traced it back and I think I found the problem.  The save function in the Items Controller, doesn't call parse_decimals the same way you do in your unit test.  It calls the function as:
`'unit_price' => parse_decimals($this->input->post('unit_price')),`

your unit test calls
`$decimals = parse_decimals(2.22, 2);

which passes, but it doesn't test the case in which parse_decimals is called without the precision variable.  This case is allowed, but parse_decimals then doesn't do any rounding.

I'm guessing that we should modify the call in the Items Controller to match app_config->currency_decimals since unit_price and cost_price are both currency values.  Unit testing is not supposed to be pulling from the database, since that's an integration testing job, but we may want to also unit_test parse_decimals with a null value in the precision variable.

@daN4cat @jekkos Thoughts

**Comment by objecttothis** [comment]
A little addendum.  I added the precision on my local copy and it didn't fix the issue.  I've traced it all the way back to my json_encode statement.  I'm going to look into understanding why json_encode is messing up the variable precision.

**Comment by objecttothis** [comment]
Found it: https://stackoverflow.com/questions/42981409/php7-1-json-encode-float-issue/43056278
I wish they would leave things alone.  We need to assess how this will affect other aspects of the code.  @daN4cat are we generally only using JSON for displaying success messages?  It seems like with versions of PHP starting in 7.1, they use a serialize with precision of 14 and you have to manually set it.  

In short, if I add the code:

````
		if (version_compare(phpversion(), '7.1', '>='))
		{
			ini_set( 'precision', 17 );
			ini_set( 'serialize_precision', -1 );
		}
````
before the json_encode, it works fine.  The problem is that We need a better place to add that code than right before the call, so that it affects all calls.

**Comment by jekkos** [comment]
> 
> 
> > @objecttothis Ok this is really weird, as I showed in the unit test, the decimal rounding works as it should. Can you check the code in between the parse en the display? Maybe extend the unit test to see where the formatting changes.
> 
> @jekkos, my code isn't manipulating the value, it's just passing it through as is until the json_encode. I traced it back and I think I found the problem. The save function in the Items Controller, doesn't call parse_decimals the same way you do in your unit test. It calls the function as:
> `'unit_price' => parse_decimals($this->input->post('unit_price')),`
> 
> your unit test calls
> `$decimals = parse_decimals(2.22, 2);
> 
> which passes, but it doesn't test the case in which parse_decimals is called without the precision variable. This case is allowed, but parse_decimals then doesn't do any rounding.
> 
> I'm guessing that we should modify the call in the Items Controller to match app_config->currency_decimals since unit_price and cost_price are both currency values. Unit testing is not supposed to be pulling from the database, since that's an integration testing job, but we may want to also unit_test parse_decimals with a null value in the precision variable.
> 
> @daN4cat @jekkos Thoughts

Well we could mock out the db call in this case so we can still verify behavior when we do not have any decimals defined (see the other test as an example, where I set barcode format that in the test itself).

**Comment by jekkos** [comment]
@objecttothis I already had a suspicion that it would be due to converting to json in this case, good find!

I think we should align the json behavior with what we configure in php for decimal and currency precision. Maybe the best place to configure this is in the decimal formatting code itself. We're not keeping any state in PHP beside what is processed in a request, so we don't have a lot of options otherwise.

**Comment by jekkos** [comment]
@objecttothis I guess this one can be closed?

**Comment by objecttothis** [comment]
> 
> 
> @objecttothis I guess this one can be closed?

Per your comments on the 24th, I need to make those changes in the code.  I assigned myself and slapped it onto 3.4.0

**Comment by jekkos** [comment]
Not sure if there is a decimal datatype in javascript. If we want to replicate the exact locale formatting in JSON then maybe it's better to export the field as a string instead.

**Comment by jekkos** [comment]
https://stackoverflow.com/questions/35709595/why-would-you-use-a-string-in-json-to-represent-a-decimal-number#38357877

**Comment by objecttothis** [comment]
> Not sure if there is a decimal datatype in javascript. If we want to replicate the exact locale formatting in JSON then maybe it's better to export the field as a string instead.

The only thing is that in the case of 3rd party integrations, such as API calls, the API endpoint may be expecting a decimal and we may not be able to control the endpoint code.  That said, I think the only thing regarding the locale that we may want to send is the number of decimal places. I don't think it would be difficult to wrap it in a parser to generate the JSON with the number of decimal places configured in the database.
